### PR TITLE
feature: support for ActiveSupport::MemCacheStore

### PR DESCRIPTION
**Pull request by elhu**
Hello,

This PR adds support for `ActiveSupport::Cache::MemCacheStore`, which is Rails' wrapper around `Dalli`.
Because of `Dalli` lack of support for `race_condition_ttl`, we're stuck with `MemCacheStore`... hence this PR!

### CODE REVIEW

**Comment by elhu** [comment]
Hmm, there's a compatibility issue with Rails 3.x's `ActiveSupport::Cache::MemCacheStore`...
Working a fix, I hope to ship it tomorrow.

**Comment by ktheory** [comment]
> Hmm, there's a compatibility issue with Rails 3.x's ActiveSupport::Cache::MemCacheStore...
> Working a fix, I hope to ship it tomorrow.

Thanks for working on that.

**Comment by elhu** [comment]
I've investigated this a bit, and there's an inconsistency in Rails 3.x, with `ActiveSupport::Cache::MemCacheStore`. 

When using the Rails 3.2 (`ActiveSupport::Cache::MemCacheStore` +  `MemCache` backend), the following returns 0:
```ruby
Rails.cache.delete("foo")
Rails.cache.increment("foo") # => 0
```
Whereas with every other cache stores, such as `Dalli`, it returns `nil`.

I'm not sure it's a good idea to deal with that specificity in `Rack::Attack`, but since Rails 3.2 is end-of-life, I don't think I can fix it there either.

I could create a new `MemCacheProxy` class to that effect, but it's only going to be useful for Rails 3.x projects with the `MemCache` backend... any opinion or alternative idea :) ?

**Comment by elhu** [comment]
I've addressed the issue by creating a dedicate MemCacheProxy class for the `ActiveSupport::Cache::MemCacheStore` + `MemCache` backend case.

The original reason for this PR is still solved, and now everything should work well even for Rails 3.2!

**Comment by jarthod** [comment]
+1

**Comment by ktheory** [comment]
@elhu Very cool. Didn't realize dalli didn't support `race_condition_ttl`! So I agree it's valuable to support `MemCacheStore`.

I'll review this as soon as I can (probably in the next couple weeks.)

If you can report success running this patch in a production application, that'd be helpful testing.

**Comment by elhu** [comment]
I'm on vacation this week, but I can start testing things out on a production app starting the week after that.

I'll adjust the PR if I find any issues, and report here if I don't find any.

**Comment by elhu** [comment]
I've brought my branch up to date with master, and fixed a couple more things.

I am also running the patch on a live web-app, and have no problems to report so far.

**Comment by elhu** [comment]
@ktheory did you get a chance to review this?
Once the PR is OK for you, let me know, I'll squash my commits together to avoid clutter in the master tree :smiley:.

File: gemfiles/activesupport3.2.gemfile [hunk 1]
```diff
@@ -3,6 +3,7 @@
 source "https://rubygems.org"
 
 gem "activesupport", "~> 3.2.0"
+gem "memcache-client"
```
Comment (ktheory) [inline gemfiles/activesupport3.2.gemfile]: Can you add `memcache-client` as a [development dependency](https://github.com/kickstarter/rack-attack/blob/d911a89b5ce0a28729674c34dd1bfcc791bce1b4/rack-attack.gemspec#L32-L33) instead?

This file is auto-generated from `Appraisals` file.

Comment (elhu) [inline gemfiles/activesupport3.2.gemfile]: I've addressed that. Didn't know about `appraisal`, very interesting stuff!
I've also refactored the `Rack::Attack::StoreProxy.build` method, which was getting too cluttered in my opinion (but I could move that to a separate PR if you think it's necessary).

**Comment by elhu** [comment]
I had to revert my refactoring of `Rack::Attack::StoreProxy.build`, MRI 2.0.0 and JRuby do not behave like MRI 2.1.x and 2.2.x when calling `Kernel.const_defined?` with a nested constant name.

**Comment by elhu** [comment]
@ktheory Is there anything more you would like to discuss regarding this PR?

**Comment by ktheory** [comment]
@elhu Looks good! :cake:

I'll aim to publish a new release by the end of the week.

**Comment by elhu** [comment]
Great, thanks!

**Comment by ktheory** [comment]
Just shipped v4.4.0: https://rubygems.org/gems/rack-attack/versions/4.4.0

**Comment by elhu** [comment]
Brilliant, thank you very much!
